### DISCUSSION TITLE
Compiler version

### DISCUSSION THREAD

**Discussion by jeyblu** [microsoft/onnxruntime#17202] [Other Q&A]
Which version of compilers are used to build the ONNX Runtime binaries released on the repository's [release page](https://github.com/microsoft/onnxruntime/releases)?
- Which MSVC version is used to build the Windows binaries?
- Which gcc version is used to build the Linux binaries?

What about the python [pip package](https://pypi.org/project/onnxruntime) for ONNX Runtime?
- Which MSVC version is used to build the Windows pip package?
- Which gcc version is used to build the Linux pip package?

Thanks

**Comment by snnn** [comment]
The latest release used Visual Studio 2019. The next release will use Visual Studio 2022 instead. 
GCC version is 11.

**Comment by jeyblu** [comment]
Which gcc 11 version? Is it gcc 11.4.0? Thanks

**Reply by snnn** [reply]
I think the minor version doesn't matter. It doesn't change the ABI.  I didn't keep it in record.

**Comment by jeyblu** [comment]
Do you know which assembler version was used? Command to check `as --version` Thanks

**Comment by snnn** [comment]
I didn't look deep into this. We just use what CentOS provides. Our Linux builds run in docker. Here is the docker file: https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/linux/docker/inference/x64/default/cpu/Dockerfile You can build the docker image and run the command by yourself. But, the output doesn't guarantee to be always the same.

**Comment by jeyblu** [comment]
The container built with the current [Dockerfile](https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/linux/docker/inference/x64/default/cpu/Dockerfile) has 

> gcc (GCC) 12.2.1 20221121 (Red Hat 12.2.1-7)

In the ONNX Runtime [release announcements](https://github.com/microsoft/onnxruntime/releases), in v1.15.0, it says

> At compiler level we will drop the support for
> GCC version <= 9
> Visual Studio 2019

The [documentation](https://onnxruntime.ai/docs/build/inferencing.html#build-environmentshost) on "Build ONNX Runtime for inferencing" says

> GCC 8.x and below are not supported. 

Is it possible to change the gcc requirement to at least gcc 12, which is what your current build container has? Newer versions of compiler are required to build better and more optimal code, including compiling new instructions for deep learning. For example, compiler support for AMX instructions which ONNX Runtime uses was added in [gcc 11](https://gcc.gnu.org/gcc-11/changes.html).

**Reply by snnn** [reply]
In any near future we cannot. Because we still need to support CUDA 11.x. But CUDA 11.8 only supports GCC 6.0-11.2.1. So GCC 12 is out of the picture.

**Reply by jeyblu** [reply]
Doesn't the requirement for an older gcc version apply only if you're building the ONNX Runtime CUDA execution provider? If you're building the ONNX Runtime CPU execution provider, then it would be better to use the latest gcc compiler, right? The build container in fact has gcc 12.2.1 which I guess is used for building the CPU execution provider and execution providers other than the CUDA execution provider. Is it possible to change the gcc requirement for execution providers other than the CUDA execution provider? Building FP16 AMX code for the CPU will require [gcc 13](https://gcc.gnu.org/gcc-13/changes.html). While we want to use the gcc version required for the CUDA execution provider, shouldn't we also want to use the gcc version required for the CPU execution provider and other execution providers? Thanks.

**Reply by snnn** [reply]
Where will the AMX code live? In the CPU execution provider or elsewhere?

**Reply by jeyblu** [reply]
In the CPU execution provider

**Reply by snnn** [reply]
But when we build with CUDA, we also need to build the CPU execution provider? They need to live in the same process, therefore I insist on using the same compiler for both to avoid potential conflicts.  
If you can scope the AMX code within #ifdefs, we might be able to release in ONNX Runtime's CPU packages, but not GPU packages.

**Reply by snnn** [reply]
@chenfucn , what do you think?

**Reply by jeyblu** [reply]
Building the CUDA execution provider and the CPU execution provider is expected to use different tools. For example, you wouldn't require CUDA compiler for the CPU execution provider. So using different compiler and tools for different execution providers is to be expected and unavoidable. Can you help me understand what's the reason for not willing to use the right compiler for the CPU execution provider? Thanks.

**Reply by snnn** [reply]
The CUDA execution provider also use code from the core framework. Therefore the core framework cannot have any code that GCC 8 cannot recognize.

**Reply by snnn** [reply]
Also, every C/C++ compiler comes with a runtime. GCC has gcc_s.so . Different GCC versions have different gcc_s.so. Therefore I don't want to mix them together.

**Reply by jeyblu** [reply]
Got it. As you suggested, if we scope the AMX code within #ifdef, then we can use the gcc version required for AMX for the CPU execution provider, and the gcc version required for the CUDA execution provider. We can try this. Thanks

**Reply by jeyblu** [reply]
CUDA [documentation](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#host-compiler-support-policy) states that gcc 6.x - 13.2 is supported. Does this address the reason for not using latest gcc?

**Reply by snnn** [reply]
You were referring CUDA 12.4's document. Not 11.x.

**Reply by jeyblu** [reply]
Do you have plans to update to CUDA 12.4?

**Reply by snnn** [reply]
We need to support both CUDA 11.x and CUDA 12.x , until Nvidia stop supporting CUDA 11.x.
